src/Native/System.Security.Cryptography.Native/pal_cms.cpp

```diff
@@ -103,3 +105,93 @@ extern "C" int CryptoNative_CmsGetRecipientStackFieldCount(CmsRecipientStack* re
 
     return sk_CMS_RecipientInfo_num(recipientStack);
 }
+
+extern "C" CMS_ContentInfo* CryptoNative_CmsEncryptInitialize(ASN1_OBJECT* algorithmOid, int32_t* status)
+{
+    if (algorithmOid == nullptr || status == nullptr)
+    {
+        *status = -1;
+        return nullptr;
+    }
+
+    const EVP_CIPHER* cipher = EVP_get_cipherbyobj(algorithmOid);
+
+    if (cipher == nullptr)
+    {
+        *status = 0;
+        return nullptr;
+    }
+
+    *status = 1;
+    return CMS_EnvelopedData_create(cipher);
+}
+
+static int CmsAddRecipients(CMS_ContentInfo* cms, X509Stack* certs, uint32_t flags)
+{
+    if (cms == nullptr || certs == nullptr)
+    {
+        return -1;
+    }
+
+    for (int i = 0; i < sk_X509_num(certs); i++) {
+        X509* recip = sk_X509_value(certs, i);
+        if (CMS_add1_recipient_cert(cms, recip, flags) == nullptr) {
+            CMSerr(CMS_F_CMS_ENCRYPT, CMS_R_RECIPIENT_ERROR);
+            return 0;
+        }
+    }
+
+    return 1;
+}
+
+extern "C" int CryptoNative_CmsAddSkidRecipients(CMS_ContentInfo* cms, X509Stack* skidCerts)
+{
+    return CmsAddRecipients(cms, skidCerts, CMS_PARTIAL | CMS_USE_KEYID);
```

Comment:
Assume for the sake of discussion that you need to make implicit SKID recipients work here.  Is there anything in the current design of this that you'd want to change?